lib/ClangImporter/ImportDecl.cpp

```diff
@@ -3612,6 +3629,13 @@ namespace {
           decl->getDefinition());
       assert(def && "Class template instantiation didn't have definition");
 
+      // Currently this is a relatively low number, in the future we might
+      // consider increasing it, but this should keep compile time down,
+      // especially for types that become exponentially large when
+      // instantiating.
+      if (getSpecializationDepth(def) > 8)
```

Comment:
API-wise, I think you should pass down the maxDepth in and exit early from the loop, instead of computing the full depth and then comparing (which could entail a bunch of wasted work).

If I instantiate a template twice with the same arguments, do we get to this point in the code twice or only once? If the answer is twice, then I can see even this being slow if you have a deep type being instantiated lots of times, and you're traversing the whole type (or up to a certain depth) for every instantiation to compute the depth.

Reply:
> API-wise, I think you should pass down the maxDepth in and exit early from the loop, instead of computing the full depth and then comparing (which could entail a bunch of wasted work).

Clever. I agree this would be a better implementation. I'll fix it.

> If I instantiate a template twice with the same arguments, do we get to this point in the code twice or only once? If the answer is twice, then I can see even this being slow if you have a deep type being instantiated lots of times, and you're traversing the whole type (or up to a certain depth) for every instantiation to compute the depth.

Unfortunately, we're going to get here twice. And you're right, this is probably one of the main reasons why deep templates are so problematic. I've got an SR to fix that: SR-14137. I'll also add a comment here linking to that SR, saying something along the lines of, "see if this still takes a long time once the SR is fixed."